### ISSUE TITLE
interpolated language model

### ISSUE THREAD

**Issue by lalimili6** [k2-fsa/icefall#1295]
Hi Dears,
In Kaldi when we interpolated the language model (trigram) the results were improved especially in expert cases like medical, etc.
In my experiments, the subword-based model didn't improve when interpolating lm and e2e models also didn't improve too, as reported in this [post](https://alphacephei.com/nsh/2021/03/02/wav2vec2.html).
1- what about k2? Are there any experiments and results?
2- and how can improve lm and results, especially in expert cases and words?
best regards

**Comment by marcoyang1998** [comment]
Generally, we observe nice improvements when decoding with language models. We have the following approaches for language information integration in k2:

1. [Shallow fusion](https://icefall.readthedocs.io/en/latest/decoding-with-langugage-models/shallow-fusion.html):
- You can use a neural network language model to improve the recognition accuracy on either in-domain or cross-domain test sets. This requires a sufficient amount of text data to train a good language model;
2. [LODR](https://icefall.readthedocs.io/en/latest/decoding-with-langugage-models/LODR.html):
- This is a more advanced version than shallow fusion; It further subtracts the score of a bi-gram model which is treated as an estimate of the internal language model of the E2E ASR model;
3. [Context biasing](https://k2-fsa.github.io/sherpa/onnx/hotwords/index.html?highlight=context):
- This one is more lightweight and might be more suitable for expert cases/words. You just need to give a list of the _context words_ and the model will try to boost the score for these words during decoding.

For more details, you may have a look at https://icefall.readthedocs.io/en/latest/decoding-with-langugage-models/index.html and #1204

**Comment by mohsen-goodarzi** [comment]
@marcoyang1998, thanks for your comprehensive reply.
Does any of the above methods support streaming decoding with LM?

**Comment by marcoyang1998** [comment]
Hi @mohsen-goodarzi, all methods above are supported for streaming decoding.

**Comment by mohsen-goodarzi** [comment]
> Hi @mohsen-goodarzi, all methods above are supported for streaming decoding.

Up to my knowledge, streaming_decode.py code does not support a decoding method with LM. Just `greedy_search`, `modified_beam_search`, and `fast_beam_search`.

Also the streaming_beam_search.py, unlike the `beam_search.py`, does not include a LM-compatible decoding method.
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless2/streaming_beam_search.py

Maybe I am looking wrong places. Can you please give an example or a staring point?

**Comment by marcoyang1998** [comment]
It's currently only supported for `--simulate-streaming` in `decode.py`, please have a look at https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/RESULTS.md#streaming-zipformer-transducer-pruned-stateless-transducer--streaming-zipformer.

It should work equally well for `streaming_decode.py`, the only thing to be done is to add a decoding method similar to https://github.com/k2-fsa/icefall/blob/52c24df61da3d04a6fdcab32d5615c394951279b/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L2687.
